examples/ex29p_proposed.cpp

```diff
@@ -0,0 +1,540 @@
+   {
+      mesh->UniformRefinement();
+   }
+
+   // 5. Define a parallel mesh by a partitioning of the serial mesh. Refine
+   //    this mesh further in parallel to increase the resolution (1 time by
+   //    default, or specified on the command line with -rp). Once the parallel
+   //    mesh is defined, the serial mesh can be deleted.
+   ParMesh pmesh(MPI_COMM_WORLD, *mesh);
+   delete mesh;
+   for (int lev = 0; lev < par_ref_levels; lev++)
+   {
+      pmesh.UniformRefinement();
+   }
+   pmesh.ReorientTetMesh();
+
+   // 6. Define a parallel finite element space on the parallel mesh. Here we
+   //    use the Nedelec finite elements of the specified order.
+   FiniteElementCollection *fec = NULL;
+   if (dim == 1)
+   {
+      fec = new ND_R1D_FECollection(order, dim);
+   }
+   else if (dim == 2)
+   {
+      fec = new ND_R2D_FECollection(order, dim);
+   }
+   else
+   {
+      fec = new ND_FECollection(order, dim);
+   }
+   ParFiniteElementSpace fespace(&pmesh, fec);
+   HYPRE_Int size = fespace.GlobalTrueVSize();
+   mfem::out << "Number of H(Curl) unknowns: " << size << endl;
+
+   // 7. Determine the list of true (i.e. parallel conforming) essential
+   //    boundary dofs. In this example, the boundary conditions are defined
+   //    by marking all the boundary attributes from the mesh as essential
+   //    (Dirichlet) and converting them to a list of true dofs.
+   Array<int> ess_tdof_list;
+   if (pmesh.bdr_attributes.Size())
+   {
+      Array<int> ess_bdr(pmesh.bdr_attributes.Max());
+      ess_bdr = 1;
+      fespace.GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
+   }
+
+   // 8. Set up the parallel linear form b(.) which corresponds to the
+   //    right-hand side of the FEM linear system, which in this case is
+   //    (f,phi_i) where f is given by the function f_exact and phi_i are the
+   //    basis functions in the finite element fespace.
+   VectorFunctionCoefficient f(3, f_exact);
+   ParLinearForm b(&fespace);
+   b.AddDomainIntegrator(new VectorFEDomainLFIntegrator(f));
+   b.Assemble();
+
+   // 9. Define the solution vector x as a parallel finite element grid function
+   //    corresponding to fespace. Initialize x by projecting the exact
+   //    solution. Note that only values from the boundary edges will be used
+   //    when eliminating the non-homogeneous boundary condition to modify the
+   //    r.h.s. vector b.
+   ParGridFunction sol(&fespace);
+   VectorFunctionCoefficient E(3, E_exact);
+   VectorFunctionCoefficient CurlE(3, CurlE_exact);
+   sol.ProjectCoefficient(E);
+
+   // 10. Set up the parallel bilinear form corresponding to the EM diffusion
+   //     operator curl muinv curl + sigma I, by adding the curl-curl and the
+   //     mass domain integrators.
+   DenseMatrix sigmaMat(3);
+   sigmaMat(0,0) = 2.0; sigmaMat(1,1) = 2.0; sigmaMat(2,2) = 2.0;
+   sigmaMat(0,2) = 0.0; sigmaMat(2,0) = 0.0;
+   sigmaMat(0,1) = M_SQRT1_2; sigmaMat(1,0) = M_SQRT1_2;
```

Comment:
This is the first time I ever saw constants like `M_SQRT1_2`, but Google helped me learn what they are. Are other people usually familiar with these, or should there be a comment about the definition?

Reply:
This may be one of the more obscure ones. I can add a comment at the end of the line if that would help to clarify things.

Reply:
Yes, it can be a short comment like `// 1/sqrt(2) in cmath`